### ISSUE TITLE
Fix S3242: Rule should not suggest IEnumerable<T> when there are multiple iterations over the collection

### ISSUE THREAD

**Issue by ohadschn** [SonarSource/sonar-csharp#680]
### Description

Collection parameters in method signatures that are more derived than `IEnumerable<T>`, even when used only as enumerables, make sense. Specifically, multiple enumerations of an IEnumerable may have undesired side effects. Resharper even has a specific analyzer to catch that.

### Repro steps

The following method will trigger S3242, suggesting the use of `IEnumerable<int>` instead of `ICollection<int>`.
```cs
public static void Foo(ICollection<int> collection)
{
    foreach (var item in collection) { Console.WriteLine(item); }
    foreach (var item in collection.OrderBy(x => x)) { Console.WriteLine(item); }
}
```

### Expected behavior

S3242 should not fire because the method enumerates the collection twice and doing so on an IEnumerable can have undesired side effects.

### Actual behavior

S3242 is fired, suggesting a potentially dangerous fix.

### Related information

* SonarC# Version: SonarAnalyzer.CSharp 6.3.0.2862
* Visual Studio Version: 2017 (15.2)

**Comment by Evangelink** [comment]
Hi @ohadschn. Thanks for this feedback, we will have a look at it soon.
